### PR TITLE
TypeScript migration for js-sdk

### PR DESCRIPTION
**Pull request by gsouquet**
- Migrate realtime-callbacks to TypeScript
- Migrate errors to TypeScript
- Migrate AutoDiscovery to TypeScript

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/realtime-callbacks.ts [hunk 1]
```diff
@@ -31,17 +31,17 @@ import { logger } from './logger';
 const TIMER_CHECK_PERIOD_MS = 1000;
 
 // counter, for making up ids to return from setTimeout
-let _count = 0;
+let count = 0;
 
 // the key for our callback with the real global.setTimeout
-let _realCallbackKey;
+let realCallbackKey;
```
Comment (t3chguy) [inline src/realtime-callbacks.ts:37]: type?

File: src/realtime-callbacks.ts [hunk 2]
```diff
@@ -31,17 +31,17 @@ import { logger } from './logger';
...
 
 // a sorted list of the callbacks to be run.
 // each is an object with keys [runAt, func, params, key].
-const _callbackList = [];
+const callbackList = [];
```
Comment (t3chguy) [inline src/realtime-callbacks.ts:41]: type?

File: src/realtime-callbacks.ts [hunk 3]
```diff
@@ -182,7 +182,7 @@ function _runCallbacks() {
  * returns the index of the last element for which func returns
  * greater than zero, or array.length if no such element exists.
  */
-function binarySearch(array, func) {
+function binarySearch<T>(array: T[], func: (T) => number): number {
```
Comment (t3chguy) [inline src/realtime-callbacks.ts:185]:
```suggestion
function binarySearch<T>(array: T[], func: (v: T) => number): number {
```

File: src/autodiscovery.ts [hunk 4]
```diff
@@ -495,8 +416,9 @@ export class AutoDiscovery {
      * @return {Promise<object>} Resolves to the returned state.
      * @private
      */
-    static async _fetchWellKnownObject(url) {
+    private static async fetchWellKnownObject(url: string): Promise<IWellKnownConfig> { // TODO: TYPES
```
Comment (t3chguy) [inline src/autodiscovery.ts:419]: what's the todo here for, looks typed to me?

**Review by t3chguy — APPROVED** [review]
Great, thanks

File: src/autodiscovery.ts (hunk 4 above)
Comment (gsouquet) [reply src/autodiscovery.ts]: Thanks for the heads up! Fixed now

**Pull request merged**